ports/servoshell/rendering_context.rs

```diff
@@ -0,0 +1,297 @@
+        let context = &mut self.0.context.borrow_mut();
+        let surface = device
+            .unbind_surface_from_context(context)
+            .unwrap()
+            .unwrap();
+        let surface = f(device, surface);
+        device.bind_surface_to_context(context, surface).unwrap();
+    }
+
+    #[allow(dead_code)]
+    pub fn connection(&self) -> Connection {
+        let device = &self.0.device.borrow();
+        device.connection()
+    }
+
+    #[allow(dead_code)]
+    pub fn adapter(&self) -> Adapter {
+        let device = &self.0.device.borrow();
+        device.adapter()
+    }
+
+    #[allow(dead_code)]
+    pub fn native_context(&self) -> NativeContext {
+        let device = &self.0.device.borrow();
+        let context = &self.0.context.borrow();
+        device.native_context(context)
+    }
+
+    #[allow(dead_code)]
+    pub fn native_device(&self) -> NativeDevice {
+        let device = &self.0.device.borrow();
+        device.native_device()
+    }
+
+    #[allow(dead_code)]
+    pub fn context_attributes(&self) -> ContextAttributes {
+        let device = &self.0.device.borrow();
+        let context = &self.0.context.borrow();
+        let descriptor = &device.context_descriptor(context);
+        device.context_descriptor_attributes(descriptor)
+    }
+
+    pub fn context_surface_info(&self) -> Result<Option<SurfaceInfo>, Error> {
+        let device = &self.0.device.borrow();
+        let context = &self.0.context.borrow();
+        device.context_surface_info(context)
+    }
+
+    #[allow(dead_code)]
+    pub fn surface_info(&self, surface: &Surface) -> SurfaceInfo {
+        let device = &self.0.device.borrow();
+        device.surface_info(surface)
+    }
+
+    #[allow(dead_code)]
+    pub fn surface_texture_object(&self, surface: &SurfaceTexture) -> u32 {
+        let device = &self.0.device.borrow();
+        device.surface_texture_object(surface)
+    }
+
+    #[allow(dead_code)]
+    pub fn get_proc_address(&self, name: &str) -> *const c_void {
+        let device = &self.0.device.borrow();
+        let context = &self.0.context.borrow();
+        device.get_proc_address(context, name)
+    }
+
+    #[allow(dead_code)]
+    pub fn unbind_native_surface_from_context(&self) -> Result<(), Error> {
+        let device = self.0.device.borrow_mut();
+        let mut context = self.0.context.borrow_mut();
+        let mut surface = device.unbind_surface_from_context(&mut context)?.unwrap();
+        device.destroy_surface(&mut context, &mut surface)?;
+        Ok(())
+    }
+}
+
+impl RenderingContext for SurfmanRenderingContext {
+    fn context(&self) -> Ref<'_, Context> {
+        self.0.context.borrow()
+    }
+
+    fn context_mut(&self) -> RefMut<'_, Context> {
+        self.0.context.borrow_mut()
+    }
+
+    fn device(&self) -> Ref<'_, Device> {
+        self.0.device.borrow()
+    }
+
+    fn device_mut(&self) -> RefMut<'_, Device> {
+        self.0.device.borrow_mut()
+    }
```

Comment:
Could these just return & and &mut. I don't think other context will all use RefCell.